File: PyEngineering/14_concurrent_cache/14_concurrent_cache_solution.py
"""
14 · Concurrent cache — solution.

See `14_concurrent_cache_explanation.py` for the full spec, rationale, and
acceptance criteria. This file is the complete, idiomatic implementation.

DESIGN DECISION: threading.Lock, not asyncio, and why
--------------------------------------------------------
See the module docstring in the explanation file for the full argument;
in short, this component is called from synchronous code paths with no
event loop, so a `threading.Lock`-protected structure (usable from any
thread with no ownership/loop constraints) is the correct default, not an
`asyncio.Lock` (which only works correctly awaited from within its owning
event loop).

DESIGN DECISION: hand-rolled single-flight via a per-key Event
-------------------------------------------------------------------
The single-flight mechanism is a small state machine per missing key:

1. The first thread to miss registers a `_SingleFlight` placeholder in
   `self._inflight[key]` **while still holding the structural lock** —
   this is the "reservation" that makes every subsequent racing thread
   see the placeholder instead of also deciding it's the one to compute.
2. That thread releases the lock, then calls `compute()` with no lock
   held at all — so hits and misses on *other* keys, and even `get()`
   calls on this key... no, `get()` on the still-inflight key correctly
   returns `None` per the spec (it must not block), which is naturally
   true here since nothing is written to `self._entries` until step 3.
3. Every other thread that reaches step "check inflight" while the
   computation is running gets the *same* `_SingleFlight` object, releases
   the lock, and blocks on its `threading.Event` outside the lock — so
   N-1 threads are simply parked on an Event, consuming no CPU, holding no
   lock, for the duration of the 1 real computation.
4. The computing thread stores the result (or exception) on the
   `_SingleFlight` object, removes it from `self._inflight`, writes the
   cache entry (success only), and calls `event.set()` — waking every
   parked thread at once, each of which reads the shared result/exception
   off the same object.

The critical invariant: the lock is held only for O(1) dict operations
(check entries, check/register inflight, on completion swap inflight for
a cache entry) — never across `compute()` itself and never across an
Event `.wait()`. That's what keeps hits, and misses on unrelated keys,
fully concurrent with a slow miss in progress on one key.
"""

from __future__ import annotations

import threading
import time
from collections.abc import Callable
from typing import Generic, TypeVar

K = TypeVar("K")
V = TypeVar("V")


class _SingleFlight(Generic[V]):
    """One in-flight computation's shared result slot + completion signal.

    Exactly one thread (the "leader") calls `compute()` and calls
    `resolve`/`fail`; every other thread ("follower") blocks on `done`
    and then reads `value`/`error`.
    """

    __slots__ = ("done", "value", "error")

    def __init__(self) -> None:
        self.done = threading.Event()
        self.value: V | None = None
        self.error: BaseException | None = None

    def resolve(self, value: V) -> None:
        self.value = value
        self.done.set()

    def fail(self, error: BaseException) -> None:
        self.error = error
        self.done.set()

    def wait_result(self) -> V:
        self.done.wait()
        if self.error is not None:
            raise self.error
        # mypy can't see that done implies exactly one of value/error is
        # set; value is only None here if the leader legitimately computed
        # None, which is a valid cached value.
        return self.value  # type: ignore[return-value]


class ConcurrentCache(Generic[K, V]):
    """Thread-safe cache: TTL eviction (lazy + optional sweep) + single-flight."""

    def __init__(
        self,
        *,
        ttl_seconds: float | None = None,
        sweep_interval_seconds: float | None = None,
        now: Callable[[], float] = time.monotonic,
    ) -> None:
        self._ttl = ttl_seconds
        self._now = now
        # Step 1: one lock protects both the entries table and the
        # in-flight registry. Both are plain dict operations (fast,
        # non-blocking work) — the lock is never held across compute() or
        # an Event.wait(), so this single lock does not become a
        # throughput bottleneck under concurrent misses on different keys.
        self._lock = threading.Lock()
        self._entries: dict[K, tuple[V, float | None]] = {}  # value, expires_at
        self._inflight: dict[K, _SingleFlight[V]] = {}

        self._sweep_thread: threading.Thread | None = None
        self._sweep_stop = threading.Event()
        if sweep_interval_seconds is not None:
            self._sweep_thread = threading.Thread(
                target=self._sweep_loop,
                args=(sweep_interval_seconds,),
                daemon=True,  # never blocks process exit even if close()
                # is forgotten, though close() is still the correct way
                # to stop it deterministically (see close()).
                name="ConcurrentCache-sweep",
            )
            self._sweep_thread.start()

    def _is_expired(self, expires_at: float | None) -> bool:
        return expires_at is not None and self._now() >= expires_at

    def get(self, key: K) -> V | None:
        """Read-only lookup: never computes, never blocks on an in-flight
        computation for this key (returns None if one is in progress)."""
        with self._lock:
            entry = self._entries.get(key)
            if entry is None:
                return None
            value, expires_at = entry
            if self._is_expired(expires_at):
                del self._entries[key]
                return None
            return value

    def get_or_compute(self, key: K, compute: Callable[[], V]) -> V:
        """Return the cached value, or compute it exactly once under
        concurrent callers racing on the same missing key."""
        # Fast path + reservation, both under the lock so they're atomic
        # with respect to every other thread calling get_or_compute.
        with self._lock:
            entry = self._entries.get(key)
            if entry is not None:
                value, expires_at = entry
                if not self._is_expired(expires_at):
                    return value
                del self._entries[key]  # lazily evict, fall through to miss

            existing = self._inflight.get(key)
            if existing is not None:
                # A follower: someone else is already computing this key.
                is_leader = False
                flight = existing
            else:
                # The leader: reserve the key for this thread BEFORE
                # releasing the lock, so every thread that arrives after
                # this point (however soon) sees `existing is not None`
                # above and becomes a follower instead of a second leader.
                is_leader = True
                flight = _SingleFlight()
                self._inflight[key] = flight

        if not is_leader:
            # Block outside the lock — a follower parked here holds no
            # lock and consumes no CPU; unrelated cache traffic proceeds
            # freely while this waits.
            return flight.wait_result()

        # Leader path: call the caller-supplied compute() with NO lock
        # held at all. This is the single most important line in the
        # file — holding self._lock here would serialize every other
        # thread's cache access (hits included) behind this one call.
        try:
            value = compute()
        except BaseException as exc:  # noqa: BLE001 - must propagate ANY
            # compute() failure to every follower, not just Exception
            # subclasses, then re-raise it ourselves too.
            with self._lock:
                del self._inflight[key]
            flight.fail(exc)
            raise
        else:
            expires_at = None if self._ttl is None else self._now() + self._ttl
            with self._lock:
                self._entries[key] = (value, expires_at)
                del self._inflight[key]
            flight.resolve(value)
            return value

    def invalidate(self, key: K) -> None:
        with self._lock:
            self._entries.pop(key, None)

    def __len__(self) -> int:
        with self._lock:
            return len(self._entries)

    def _sweep_loop(self, interval: float) -> None:
        # Runs on the daemon sweep thread. wait() with a timeout doubles
        # as both the sleep and the stop signal — close() sets the event
        # to end this loop immediately instead of waiting out the full
        # interval.
        while not self._sweep_stop.wait(timeout=interval):
            with self._lock:
                expired = [
                    k for k, (_, exp) in self._entries.items() if self._is_expired(exp)
                ]
                for k in expired:
                    del self._entries[k]

    def close(self) -> None:
        """Stop the background sweep thread, if any, deterministically.

        Idempotent: safe to call more than once.
        """
        if self._sweep_thread is not None:
            self._sweep_stop.set()
            self._sweep_thread.join()
            self._sweep_thread = None


# ---------------------------------------------------------------------------
# Best practices demonstrated here
# ---------------------------------------------------------------------------
# - Never hold a lock across a caller-supplied callback (`compute()`) or a
#   blocking wait (`Event.wait()`) — both must happen with the lock
#   released, or the entire cache serializes behind the slowest miss.
# - "Reserve, then release, then do the slow thing" is the general pattern
#   for hand-rolled single-flight/de-duplication: the reservation must be
#   atomic with the check that decides "am I the leader," which is exactly
#   what doing both under one lock acquisition guarantees.
# - Don't cache failures. A cache is a performance optimization over a
#   presumed-idempotent, presumed-eventually-successful computation; an
#   exception is information about *this* attempt, not a fact about the
#   key that deserves to be memoized.
# - Inject the clock (`now: Callable[[], float]`) for TTL tests — a real
#   `time.sleep()`-based TTL test is slow and flaky under CI load; a fake
#   clock makes expiry assertions instant and deterministic.
# - Make the background thread a daemon AND give `close()` a real, joined
#   shutdown path — belt and suspenders. The daemon flag prevents an
#   at-exit hang if a caller forgets `close()`; the explicit join makes
#   "no leftover thread" a testable, deterministic guarantee for callers
#   who do the right thing.
#
# Alternative approaches
# -----------------------
# - Per-key locks (`dict[K, threading.Lock]`) instead of one structural
#   lock — reduces contention between unrelated keys' reservation step,
#   but that step is already O(1) dict work under the GIL, so it rarely
#   matters in practice; it adds real complexity (locks-for-locks: you now
#   need to synchronize creation/cleanup of the per-key lock dict itself).
# - `functools.lru_cache` / `cachetools.TTLCache` for the non-single-flight
#   subset of this problem — neither offers single-flight de-duplication
#   or a background sweep out of the box; both are good choices when
#   thundering-herd protection genuinely doesn't matter for the workload.
# - A read-write lock (no stdlib primitive; would need a third-party
#   package or a hand-rolled one) to let concurrent `get()` reads proceed
#   without even the brief exclusive lock used here — usually not worth
#   the complexity given how short the critical sections already are.
#
# Testing/benchmarking notes
# ----------------------------
# - The single-flight property is only provable under genuine concurrent
#   contention: use `threading.Barrier` to line up N threads and release
#   them at the same instant, then assert `compute` was called exactly
#   once — sequential calls to `get_or_compute` (even in a tight loop)
#   cannot expose a broken single-flight implementation, because the race
#   window is a handful of instructions wide.
# - TTL tests should control time explicitly via the injected `now`
#   callable rather than `time.sleep`, both for speed and to eliminate
#   scheduler-jitter flakiness near the expiry boundary.

File: GoEngineering/14_concurrent_cache/solution/14_concurrent_cache_solution.go
/*
Problem 14 — Concurrent Cache (reference solution)

See explanation/14_concurrent_cache_explanation.go for the full spec and
rationale. This file implements it.

Design summary:
  - Two storage backends (StoreMutex, StoreSyncMap) behind one internal
    `store[K, V]` interface, selected at construction. Both hold entry[V]
    values (value + expiry deadline). Benchmarks at the bottom of the test
    file compare them directly instead of asserting a winner in prose.
  - Single-flight is entirely separate from storage: an `inflight
    map[K]*call[V]` guarded by its own mutex. The first goroutine to miss
    on a key creates the *call[V], releases the inflight lock, runs the
    loader OUTSIDE any lock, then stores the result and closes call.done.
    Every other concurrently-missing goroutine for the same key finds the
    existing *call[V] under the same brief lock and waits on call.done —
    so exactly one loader invocation happens per stampede.
  - Lazy expiry-on-read is the correctness backstop; the background
    sweeper is the memory-bounding mechanism. Either alone is insufficient
    per the explanation file's pitfalls section.
  - Close() stops the sweeper deterministically: closes a stop channel and
    waits on a sync.WaitGroup for the sweeper goroutine to actually exit,
    so Close is synchronous rather than "requested."
*/
package cache

import (
	"context"
	"sync"
	"time"
)

// Loader loads the value for key on a cache miss.
type Loader[K comparable, V any] func(ctx context.Context, key K) (V, error)

// Store selects the internal storage strategy.
type Store string

const (
	// StoreMutex backs the cache with a plain map[K]entry[V] guarded by a
	// sync.Mutex. Default when Options.Store is empty; faster than
	// StoreSyncMap for most cache workloads (read+write mixed on a shared
	// key set) per the stdlib's own sync.Map guidance.
	StoreMutex Store = "mutex"
	// StoreSyncMap backs the cache with sync.Map — optimized for
	// mostly-write-once/read-many or disjoint-key-per-goroutine access
	// patterns, not a general drop-in replacement for a mutex+map cache.
	StoreSyncMap Store = "syncmap"
)

// Options configures a new Cache.
type Options struct {
	// TTL is how long an entry stays valid after being set. Required, > 0.
	TTL time.Duration
	// SweepInterval is how often the background sweeper scans for and
	// removes expired entries. Required, > 0.
	SweepInterval time.Duration
	// Store selects StoreMutex or StoreSyncMap. Zero value defaults to
	// StoreMutex.
	Store Store
}

// entry is one stored value plus its expiry deadline.
type entry[V any] struct {
	value     V
	expiresAt time.Time
}

func (e entry[V]) expired(now time.Time) bool {
	return now.After(e.expiresAt)
}

// store is the storage abstraction the two backends implement. Kept
// intentionally tiny — exactly what Cache needs, nothing mirroring a
// general-purpose map API.
type store[K comparable, V any] interface {
	load(key K) (entry[V], bool)
	store(key K, e entry[V])
	delete(key K)
	// sweep removes every entry for which expired(now) is true and
	// returns how many were removed (used by tests; harmless in prod).
	sweep(now time.Time) int
	// len reports the number of currently live (non-expired) entries.
	len(now time.Time) int
}

// mutexStore is a map[K]entry[V] guarded by a sync.Mutex.
type mutexStore[K comparable, V any] struct {
	mu sync.Mutex
	m  map[K]entry[V]
}

func newMutexStore[K comparable, V any]() *mutexStore[K, V] {
	return &mutexStore[K, V]{m: make(map[K]entry[V])}
}

func (s *mutexStore[K, V]) load(key K) (entry[V], bool) {
	s.mu.Lock()
	defer s.mu.Unlock()
	e, ok := s.m[key]
	return e, ok
}

func (s *mutexStore[K, V]) store(key K, e entry[V]) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.m[key] = e
}

func (s *mutexStore[K, V]) delete(key K) {
	s.mu.Lock()
	defer s.mu.Unlock()
	delete(s.m, key)
}

func (s *mutexStore[K, V]) sweep(now time.Time) int {
	s.mu.Lock()
	defer s.mu.Unlock()
	n := 0
	for k, e := range s.m {
		if e.expired(now) {
			delete(s.m, k)
			n++
		}
	}
	return n
}

func (s *mutexStore[K, V]) len(now time.Time) int {
	s.mu.Lock()
	defer s.mu.Unlock()
	n := 0
	for _, e := range s.m {
		if !e.expired(now) {
			n++
		}
	}
	return n
}

// syncMapStore wraps sync.Map. Len/sweep necessarily do a full Range scan —
// sync.Map has no size accounting, which is itself part of the trade-off
// the explanation file asks you to weigh.
type syncMapStore[K comparable, V any] struct {
	m sync.Map // K -> entry[V]
}

func newSyncMapStore[K comparable, V any]() *syncMapStore[K, V] {
	return &syncMapStore[K, V]{}
}

func (s *syncMapStore[K, V]) load(key K) (entry[V], bool) {
	v, ok := s.m.Load(key)
	if !ok {
		return entry[V]{}, false
	}
	return v.(entry[V]), true
}

func (s *syncMapStore[K, V]) store(key K, e entry[V]) {
	s.m.Store(key, e)
}

func (s *syncMapStore[K, V]) delete(key K) {
	s.m.Delete(key)
}

func (s *syncMapStore[K, V]) sweep(now time.Time) int {
	n := 0
	s.m.Range(func(k, v any) bool {
		if v.(entry[V]).expired(now) {
			s.m.Delete(k)
			n++
		}
		return true
	})
	return n
}

func (s *syncMapStore[K, V]) len(now time.Time) int {
	n := 0
	s.m.Range(func(_, v any) bool {
		if !v.(entry[V]).expired(now) {
			n++
		}
		return true
	})
	return n
}

// call is one in-flight single-flight load for a given key: the first
// goroutine to miss on a key creates one of these, runs the load, and
// every other concurrently-missing goroutine for the SAME key waits on
// done instead of calling load itself — the classic "future" shape.
type call[V any] struct {
	done  chan struct{}
	value V
	err   error
}

// Cache is a generic, concurrency-safe, TTL-expiring cache with
// single-flight loading. The zero value is not usable; construct with New.
type Cache[K comparable, V any] struct {
	ttl   time.Duration
	store store[K, V]

	inflightMu sync.Mutex
	inflight   map[K]*call[V]

	stop     chan struct{}
	stopOnce sync.Once
	wg       sync.WaitGroup
}

// New constructs a ready-to-use Cache and starts its background sweeper
// goroutine. Panics if opts.TTL <= 0 or opts.SweepInterval <= 0.
func New[K comparable, V any](opts Options) *Cache[K, V] {
	if opts.TTL <= 0 {
		panic("cache: Options.TTL must be > 0")
	}
	if opts.SweepInterval <= 0 {
		panic("cache: Options.SweepInterval must be > 0")
	}

	var st store[K, V]
	switch opts.Store {
	case StoreSyncMap:
		st = newSyncMapStore[K, V]()
	case StoreMutex, "":
		st = newMutexStore[K, V]()
	default:
		panic("cache: unknown Store: " + string(opts.Store))
	}

	c := &Cache[K, V]{
		ttl:      opts.TTL,
		store:    st,
		inflight: make(map[K]*call[V]),
		stop:     make(chan struct{}),
	}

	c.wg.Add(1)
	go c.sweepLoop(opts.SweepInterval)

	return c
}

// sweepLoop periodically removes expired entries until Close stops it.
// Each sweep only holds the store's internal lock for the duration of the
// scan itself (see mutexStore.sweep), never for the full SweepInterval.
func (c *Cache[K, V]) sweepLoop(interval time.Duration) {
	defer c.wg.Done()
	t := time.NewTicker(interval)
	defer t.Stop()
	for {
		select {
		case <-c.stop:
			return
		case now := <-t.C:
			c.store.sweep(now)
		}
	}
}

// Get returns the cached value for key if present and unexpired. On a
// miss, it calls load — collapsing concurrent misses for the same key into
// exactly one loader call (single-flight) — stores the result on success
// with a fresh TTL, and returns it to every waiter. A loader error is
// returned to every waiter and is never cached.
func (c *Cache[K, V]) Get(ctx context.Context, key K, load Loader[K, V]) (V, error) {
	now := time.Now()

	// Fast path: cached and not expired.
	if e, ok := c.store.load(key); ok && !e.expired(now) {
		return e.value, nil
	}

	// Slow path: join-or-create the in-flight call for this key. This
	// whole check-and-maybe-create step is one critical section under the
	// inflight mutex — otherwise two goroutines could both observe "no
	// in-flight call" and both start their own loader, defeating
	// single-flight.
	c.inflightMu.Lock()
	if existing, ok := c.inflight[key]; ok {
		c.inflightMu.Unlock()
		return c.wait(ctx, existing)
	}
	cl := &call[V]{done: make(chan struct{})}
	c.inflight[key] = cl
	c.inflightMu.Unlock()

	// Run the loader OUTSIDE any lock: it may be slow (network/DB call)
	// and must not block Get for unrelated keys, nor hold the inflight
	// lock while other keys try to join-or-create their own calls.
	value, err := load(ctx, key)
	cl.value, cl.err = value, err

	if err == nil {
		c.store.store(key, entry[V]{value: value, expiresAt: time.Now().Add(c.ttl)})
	}

	c.inflightMu.Lock()
	// Only remove our own call — a concurrent Get for the same key that
	// missed again after this one completed (rare, but possible right at
	// the boundary) would have created a NEW *call[V] under the lock, and
	// we must not delete that one.
	if c.inflight[key] == cl {
		delete(c.inflight, key)
	}
	c.inflightMu.Unlock()

	close(cl.done)
	return value, err
}

// wait blocks until the in-flight call cl completes or ctx is canceled.
// Canceling ctx only stops THIS waiter from waiting — it does not cancel
// the load for other waiters, who did not ask to cancel it.
func (c *Cache[K, V]) wait(ctx context.Context, cl *call[V]) (V, error) {
	select {
	case <-cl.done:
		return cl.value, cl.err
	case <-ctx.Done():
		var zero V
		return zero, ctx.Err()
	}
}

// Set unconditionally stores value for key with a fresh TTL, bypassing the
// loader path entirely.
func (c *Cache[K, V]) Set(key K, value V) {
	c.store.store(key, entry[V]{value: value, expiresAt: time.Now().Add(c.ttl)})
}

// Delete removes key immediately, regardless of TTL.
func (c *Cache[K, V]) Delete(key K) {
	c.store.delete(key)
}

// Len reports the number of currently live (non-expired) entries. For the
// StoreSyncMap backend this is an O(n) full-Range scan — sync.Map has no
// cheap size accounting, which is itself part of the mutex-vs-sync.Map
// trade-off.
func (c *Cache[K, V]) Len() int {
	return c.store.len(time.Now())
}

// Close stops the background sweeper goroutine and waits for it to
// actually exit before returning (synchronous, not "requested"). Safe to
// call more than once. Get/Set/Delete remain safe to call afterward — they
// simply no longer benefit from background sweeping; the lazy
// expiry-on-read check in Get still applies.
func (c *Cache[K, V]) Close() {
	c.stopOnce.Do(func() {
		close(c.stop)
	})
	c.wg.Wait()
}

/*
BEST PRACTICES DEMONSTRATED

  - Single-flight bookkeeping (inflightMu/inflight) is a SEPARATE lock from
    storage, and the loader itself runs with NO lock held — the only locked
    sections are the O(1) map operations bracketing it. This is what keeps
    one slow loader from blocking Gets for unrelated keys.
  - The storage backend is an interface (`store[K, V]`) with two
    implementations selected at construction, not a runtime `if` scattered
    through every method — makes both variants independently benchmarkable
    and keeps Cache's own logic backend-agnostic.
  - Lazy expiry-on-read (the `!e.expired(now)` check in the Get fast path)
    plus a background sweeper are BOTH present: the sweeper bounds memory
    over time, the lazy check is the correctness backstop that catches
    anything the sweeper hasn't reached yet (relevant whenever
    SweepInterval > TTL, a legitimate configuration).
  - Close() is synchronous: it waits on a sync.WaitGroup for the sweeper
    goroutine to actually observe the stop channel and return, rather than
    just signaling and hoping — callers can rely on "no goroutine owned by
    this cache is running" being true immediately after Close returns.
  - A loader error is never written to the store — only the success path
    calls c.store.store — so a transient failure doesn't poison the cache
    for a full TTL.

ALTERNATIVE APPROACHES

  - golang.org/x/sync/singleflight provides Group.Do with the same
    join-or-create shape implemented here, plus a Forget method for
    explicitly evicting an in-flight (or just-completed) call — worth
    reaching for in production code once you understand this mechanism;
    reimplementing it here is purely so the mechanism isn't a black box.
  - sync.Map's LoadOrStore could implement the inflight join-or-create step
    atomically without a separate mutex — but LoadOrStore always
    constructs its second argument eagerly even on the Load path, so
    you'd allocate a fresh *call[V] on every single Get, hit or miss; the
    explicit mutex here only allocates a *call[V] when actually needed.
  - An LRU/LFU size-bounded eviction policy layered on top of TTL (stretch
    goal in the explanation file) is a common production addition when
    memory must be bounded by entry count, not just by time.
  - A "loading placeholder" stored directly in the main store (instead of a
    separate inflight map) is another common single-flight shape, but
    requires every store method to understand two entry states (loading vs
    loaded) instead of keeping that state cleanly out of the storage layer.
*/
